File: 模板/python/KMP.py
def prefix_function(s):
    n = len(s)
    pi = [0] * n
    for i in range(1, n):
        j = pi[i - 1]
        while j > 0 and s[i] != s[j]:
            j = pi[j - 1]
        if s[i] == s[j]:
            j += 1
        pi[i] = j
    return pi

def kmp_automaton(s):
    n = len(s)
    s = [ord(i) - 97 for i in s]
    nxt = [[0] * 26 for _ in range(n+1)]
    j = 0
    for i in range(1, n+1):
        j = nxt[j][s[i-1]]
        nxt[i-1][s[i-1]] = i
        nxt[i] = deepcopy(nxt[j])
    return nxt

File: 模板/cpp/KMP.cpp
vector<int> prefix_function(string& s) {
  int n = (int)s.length();
  vector<int> pi(n);
  for (int i = 1; i < n; i++) {
    int j = pi[i - 1];
    while (j > 0 && s[i] != s[j]) j = pi[j - 1];
    if (s[i] == s[j]) j++;
    pi[i] = j;
  }
  return pi;
}

vector<vector<int>> kmp_automaton(string& s) {
    int n = s.size();
    vector<vector<int>> nxt(n+1, vector<int>(26));
    for (int i = 1, j = 0; i <= n; i++) {
        j = nxt[j][s[i-1] - 'a'];
        nxt[i-1][s[i-1] - 'a'] = i;
        nxt[i] = vector<int>(nxt[j].begin(), nxt[j].end());
    }
    return nxt;
}

vector<int> z_function(string s) {
  int n = (int)s.length();
  vector<int> z(n);
  for (int i = 1, l = 0, r = 0; i < n; ++i) {
    if (i <= r && z[i - l] < r - i + 1) {
      z[i] = z[i - l];
    } else {
      z[i] = max(0, r - i + 1);
      while (i + z[i] < n && s[z[i]] == s[i + z[i]]) ++z[i];
    }
    if (i + z[i] - 1 > r) l = i, r = i + z[i] - 1;
  }
  return z;
}
